### PR TITLE
Fix ApplyStashedOp for DataStore and DDS Attach Messages

### PR DESCRIPTION
**Pull request by anthony-murphy**
This change adds support to applyStashedOp for DataStore and DDS attach messages while in a detached state, and unifies that with how applyStashedOp works in the attached state. 

Related [AB#7452](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/7452)

### CODE REVIEW

File: packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts [hunk 1]
```diff
@@ -95,171 +94,184 @@ describeCompat(
...
+				rootOfDataStore1.set(dataStore2Key, dataStore2.handle);
+
+				// serialize and rehydrate
+				const snapshotTree = container.serialize();
+
+				const loader = createTestLoader(provider);
+
+				const rehydratedContainer =
+					await loader.rehydrateDetachedContainerFromSnapshot(snapshotTree);
+
```
Comment (jzaffiro) [inline packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts:144]: I think you're missing the second attach attempt here, unless we don't want to check attach after rehydrate for data stores.

File: packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts [hunk 2]
```diff
@@ -95,171 +94,184 @@ describeCompat(
...
+				if (attachAfterRehydrate) {
+					await rehydratedContainer.attach(provider.driver.createCreateNewRequest());
```
Comment (jzaffiro) [inline packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts:246]: Should the attach state of rehydratedContainer be validated here?

File: packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts [hunk 3]
```diff
@@ -95,171 +94,184 @@ describeCompat(
...
-			assert(dataStore2FromRC, "DataStore2 should have been serialized properly");
-			assert.strictEqual(
-				dataStore2FromRC.runtime.id,
-				dataStore2.runtime.id,
-				"DataStore2 id should match",
-			);
+				// Check for default data store
+				const entryPoint = await rehydratedContainer.getEntryPoint();
+				assert.notStrictEqual(entryPoint, undefined, "Component should exist!");
+				// Might need some other asserts, but not sure what
```
Comment (jzaffiro) [inline packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts:117]:
```suggestion
				
```
Apparently I left this in my pr :)

File: packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts [hunk 4]
```diff
@@ -5,14 +5,9 @@
 
 import { strict as assert } from "assert";
```
Comment (anthony-murphy) [inline packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts:6]: this file is easier to review by hiding whitespace, as it reuses tests with slight modifications, so they get tabbed in another level

File: packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts [hunk 5]
```diff
@@ -38,10 +33,16 @@ describeCompat(
 		};
 		const sharedStringId = "ss1Key";
 		const sharedMapId = "sm1Key";
-		async function createDetachedContainerAndGetEntryPoint(
-			loader: IHostLoader,
-			request: IRequest,
-		) {
+		async function createDetachedContainerAndGetEntryPoint(provider: ITestObjectProvider) {
```
Comment (Abe27342) [inline packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts:36]: nit: now that the failure injection (and not just the expectations) happens in this function, `createAndFailToAttachContainer` seems like a more fitting name.

File: packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts [hunk 6]
```diff
@@ -95,171 +94,191 @@ describeCompat(
...
 
-		it("Can serialize detached container", async () => {
-			const provider = getTestObjectProvider();
-			const documentId = createDocumentId();
-			const request = provider.driver.createCreateNewRequest(documentId);
-			const loader = createTestLoader(provider);
-
-			const { container } = await createDetachedContainerAndGetEntryPoint(loader, request);
+		for (const attachAfterRehydrate of [true, false]) {
+			it(`Can serialize detached container. attachAfterRehydrate: ${attachAfterRehydrate}`, async () => {
```
Comment (Abe27342) [inline packages/test/test-end-to-end-tests/src/test/serializeAfterFailedAttach.spec.ts:98]: pre-existing, but "detached" in all of these test case names seems a bit weird: they all use the helper which puts the container in an attaching state before serializing.

anyway, I think it's worth including in the test case name the fact that this one has *no* operations after the failed attach. Makes its use as a simpler variant of below tests more obvious.

File: packages/runtime/container-runtime/src/channelCollection.ts [hunk 7]
```diff
@@ -746,9 +746,32 @@ export class ChannelCollection implements IFluidDataStoreChannel, IDisposable {
...
+			parentContext: this.wrapContextForInnerChannel(id),
+			storage,
+			scope: this.parentContext.scope,
+			createSummarizerNodeFn: this.parentContext.getCreateChildSummarizerNodeFn(id, {
+				type: CreateSummarizerNodeSource.FromSummary,
+			}),
+			makeLocallyVisibleFn: () => this.makeDataStoreLocallyVisible(id),
+			snapshotTree,
+		});
+		await dataStoreContext.realize().then(async (e) => e.entryPoint.get());
```
Comment (Abe27342) [inline packages/runtime/container-runtime/src/channelCollection.ts:770]:
```suggestion
		const channel = await dataStoreContext.realize();
		await channel.entryPoint.get();
```

File: packages/runtime/container-runtime/src/channelCollection.ts [hunk 8]
```diff
@@ -746,9 +746,32 @@ export class ChannelCollection implements IFluidDataStoreChannel, IDisposable {
 	}
 
 	private async applyStashedAttachOp(message: IAttachMessage) {
-		this.pendingAttach.set(message.id, message);
-		// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
-		this.processAttachMessage({ contents: message } as ISequencedDocumentMessage, false);
+		const { id, snapshot } = message;
```
[the remark this answers is not in the document]
Comment (anthony-murphy) [reply packages/runtime/container-runtime/src/channelCollection.ts:749]: previously all the code basically treated stashed attaches as remote messages. this kind of work before detached, but in detached there are not remote messages, so this changes it behave more like a local create, which works is detached, and is more accurate for the attached case.

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**